app/models/organization.rb

```diff
@@ -45,4 +45,56 @@ def exchange_rates
       end
     end
   end
+
+  def by_customers(categories_type, period)
+    sum = categories_type == :expenses ? 'sum(abs(transactions.amount_cents))' : 'sum(transactions.amount_cents)'
+
+    selection = transactions.unscope(:order).period(period).
+      select("#{sum} as total, customers.name as cust_name, customers.id as customer_id, bank_accounts.currency as curr").
+      joins(:customer).
+      where('transactions.category_id in (?)', categories.send(categories_type).pluck(:id)).
+      group('customers.id, bank_accounts.id').map do |transaction|
+        {
+          total:         transaction.total.to_f,
+          customer_id:   transaction.customer_id,
+          customer_name: transaction.cust_name,
+          currency:      transaction.curr
+        }
+      end
+    customers = {}
+    data, ids = [], []
+    calc_to_def_currency(customers, selection)
+    calc_total_for_customer(customers, selection, data, ids)
+    data.size > 1 ? { data: data, ids: ids, currency_format: currency_format } : nil
+  end
+
+  private
+
+  def calc_to_def_currency(customers, selection)
+    selection.each do |income|
+      customers[income[:customer_id]] = income[:customer_name]
+      if income[:currency] != default_currency
+        income[:total] = Money.new(income[:total], income[:currency]).exchange_to(default_currency).cents
+        income[:currency] = default_currency
+      end
+    end
+  end
+
+  def calc_total_for_customer(customers, selection, data, ids)
+    customers.each_pair do |id, name|
+      total_for_customer = 0
+      selection.each do |income|
+        total_for_customer += income[:total] if income[:customer_id] == id
+      end
+      ids  << id
```

Comment:
@yunixon it will be better to use hash instead two arrays. And don't provide this hash as parameter. Create and return it from function. Later if you need separate arrays for ids and data you may cal `hash.keys` and `hash.values`.